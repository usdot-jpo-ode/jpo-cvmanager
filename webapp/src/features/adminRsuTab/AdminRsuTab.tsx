import React, { useEffect, useState } from 'react'
import AdminAddRsu from '../adminAddRsu/AdminAddRsu'
import AdminEditRsu, { AdminEditRsuFormType } from '../adminEditRsu/AdminEditRsu'
import AdminTable from '../../components/AdminTable'
import { IoChevronBackCircleOutline, IoRefresh } from 'react-icons/io5'
import { AiOutlinePlusCircle } from 'react-icons/ai'
import { confirmAlert } from 'react-confirm-alert'
import { Options } from '../../components/AdminDeletionOptions'
import {
  selectLoading,
  selectTableData,
  selectEditRsuRowData,

  // actions
  updateTableData,
  deleteMultipleRsus,
  deleteRsu,
  setEditRsuRowData,
} from './adminRsuTabSlice'
import { useSelector, useDispatch } from 'react-redux'

import './Admin.css'
import { AnyAction, ThunkDispatch } from '@reduxjs/toolkit'
import { RootState } from '../../store'
import { Action } from '@material-table/core'
import { Route, Routes, useLocation, useNavigate } from 'react-router-dom'
import { NotFound } from '../../pages/404'
import toast from 'react-hot-toast'
import { ContainedIconButton } from '../../styles/components/ContainedIconButton'

const getTitle = (activeTab: string) => {
  if (activeTab === undefined) {
    return 'CV Manager RSUs'
  } else if (activeTab === 'editRsu') {
    return ''
  } else if (activeTab === 'addRsu') {
    return ''
  }
  return 'Unknown'
}

const AdminRsuTab = () => {
  const dispatch: ThunkDispatch<RootState, void, AnyAction> = useDispatch()
  const navigate = useNavigate()
  const location = useLocation()

  const activeTab = location.pathname.split('/')[4]
  const title = getTitle(activeTab)

  const tableData = useSelector(selectTableData)
  const [columns] = useState([
    { title: 'Milepost', field: 'milepost', id: 0 },
    { title: 'IP Address', field: 'ip', id: 1 },
    { title: 'Primary Route', field: 'primary_route', id: 2 },
    { title: 'RSU Model', field: 'model', id: 3 },
    { title: 'Serial Number', field: 'serial_number', id: 4 },
  ])

  const loading = useSelector(selectLoading)

  const tableActions: Action<AdminEditRsuFormType>[] = [
    {
      icon: 'delete',
      tooltip: 'Delete RSU',
      position: 'row',
      onClick: (event, rowData: AdminEditRsuFormType) => {
        const buttons = [
          { label: 'Yes', onClick: () => onDelete(rowData) },
          { label: 'No', onClick: () => {} },
        ]
        const alertOptions = Options('Delete RSU', 'Are you sure you want to delete "' + rowData.ip + '"?', buttons)
        confirmAlert(alertOptions)
      },
    },
    {
      icon: 'edit',
      tooltip: 'Edit RSU',
      position: 'row',
      onClick: (event, rowData: AdminEditRsuFormType) => onEdit(rowData),
    },
    {
      tooltip: 'Remove All Selected From Organization',
      icon: 'delete',
      onClick: (event, rowData: AdminEditRsuFormType[]) => {
        const buttons = [
          { label: 'Yes', onClick: () => multiDelete(rowData) },
          { label: 'No', onClick: () => {} },
        ]
        const alertOptions = Options(
          'Delete Selected RSUs',
          'Are you sure you want to delete ' + rowData.length + ' RSUs?',
          buttons
        )
        confirmAlert(alertOptions)
      },
    },
  ]

  const onEdit = (row: AdminEditRsuFormType) => {
    dispatch(setEditRsuRowData(row))
    navigate('editRsu/' + row.ip)
  }

  const onDelete = (row: AdminEditRsuFormType) => {
    dispatch(deleteRsu({ rsu_ip: row.ip, shouldUpdateTableData: true })).then((data: any) => {
      data.payload.success
        ? toast.success('RSU Deleted Successfully')
        : toast.error('Failed to delete RSU due to error: ' + data.payload)
    })
  }

  const multiDelete = (rows: AdminEditRsuFormType[]) => {
    dispatch(deleteMultipleRsus(rows)).then((data: any) => {
      data.payload.success ? toast.success('RSUs Deleted Successfully') : toast.error(data.payload.message)
    })
  }

  return (
    <div>
      <div>
        <h3 className="panel-header" key="adminRsuTab">
          {title}
          {activeTab === undefined && [
            <>
              <ContainedIconButton
                key="plus_button"
                title="Add RSU"
                onClick={() => navigate('addRsu')}
                sx={{
                  float: 'right',
                  margin: 2,
                  mt: -0.5,
                  mr: 0,
                  ml: 0.5,
                }}
              >
                <AiOutlinePlusCircle size={20} />
              </ContainedIconButton>
              <ContainedIconButton
                key="refresh_button"
                title="Refresh RSUs"
                onClick={() => dispatch(updateTableData())}
                sx={{
                  float: 'right',
                  margin: 2,
                  mt: -0.5,
                  mr: 0,
                  ml: 0.5,
                }}
              >
                <IoRefresh size={20} />
              </ContainedIconButton>
            </>,
          ]}
        </h3>
      </div>
      <Routes>
        <Route
          path="/"
          element={
            loading === false && (
              <div className="scroll-div-tab">
                <AdminTable title={''} data={tableData} columns={columns} actions={tableActions} />
              </div>
            )
          }
        />
        <Route path="addRsu" element={<AdminAddRsu />} />
        <Route path="editRsu/:rsuIp" element={<AdminEditRsu />} />
        <Route
          path="*"
          element={
            <NotFound
              redirectRoute="/dashboard/admin/rsus"
              redirectRouteName="Admin RSU Page"
              offsetHeight={319}
              description="This page does not exist. Please return to the admin RSU page."
            />
          }
        />
      </Routes>
    </div>
  )
}

export default AdminRsuTab
