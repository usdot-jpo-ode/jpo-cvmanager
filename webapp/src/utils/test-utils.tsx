/**
 * This function searches for the every somewhat random/chaotic class name and property that cause snapshot tests to be inconsistent.
 * This current list includes MUI classes (css-*) and aria-invalid attributes.
 *
 * @ex :
 * ```
 * const { container } = render(<Component />);
 *
 * replaceChaoticIds(container);
 * ```
 *
 * @param container The HTMLElement node to search for SSR ids
 */
function replaceChaoticIds(container: HTMLElement) {
  const props = [
    {
      selector: 'class',
      updateFunc: (val: string) => val.replace(/css-[0-9a-z]*?-/g, 'css-mocked-'),
    },
  ]

  container.querySelectorAll('*').forEach((item) => {
    props.forEach((prop) => {
      if (item.getAttribute(prop.selector)) {
        item.setAttribute(prop.selector, prop.updateFunc(item.getAttribute(prop.selector)))
      }
    })
  })
  container.querySelectorAll('input[aria-invalid]').forEach((item) => {
    item.removeAttribute('aria-invalid')
  })
  return container
}

import React from 'react'
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider'
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs'
import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'

// Extend dayjs with timezone and UTC plugins
dayjs.extend(utc)
dayjs.extend(timezone)

// Mock component to set timezone
const MockLocalizationProvider = ({ children }) => {
  dayjs.tz.setDefault('America/Denver')
  return <LocalizationProvider dateAdapter={AdapterDayjs}>{children}</LocalizationProvider>
}

export { replaceChaoticIds, MockLocalizationProvider }
