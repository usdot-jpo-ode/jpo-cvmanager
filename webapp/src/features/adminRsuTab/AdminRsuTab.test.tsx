import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import AdminRsuTab from './AdminRsuTab'
import { Provider } from 'react-redux'
import { ThemeProvider } from '@mui/material'
import { testTheme } from '../../styles'
import { setupStore } from '../../store'
import { replaceChaoticIds } from '../../utils/test-utils'
import { BrowserRouter } from 'react-router-dom'

it('should take a snapshot', () => {
  const { container } = render(
    <ThemeProvider theme={testTheme}>
      <Provider store={setupStore({ adminRsuTab: { loading: false, value: { activeDiv: 'rsu_table' } } })}>
        <BrowserRouter>
          <AdminRsuTab />
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  )

  expect(replaceChaoticIds(container)).toMatchSnapshot()
})
