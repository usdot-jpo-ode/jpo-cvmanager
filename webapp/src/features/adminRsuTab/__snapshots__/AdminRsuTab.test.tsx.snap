// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<div>
  <div>
    <div
      class="scroll-div-tab"
    >
      <div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-mocked-MuiPaper-root"
          style="position: relative; font-family: sans-serif;"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-mocked-MuiToolbar-root"
          >
            
            <div
              class="MuiFormControl-root MuiTextField-root css-mocked-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-mocked-MuiInputBase-root-MuiInput-root"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-mocked-MuiInputAdornment-root"
                >
                  <span
                    class="notranslate"
                  >
                    â€‹
                  </span>
                  <svg
                    aria-hidden="true"
                    aria-label="Search"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-mocked-MuiSvgIcon-root"
                    data-mui-internal-clone-element="true"
                    data-testid="search"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                    />
                  </svg>
                </div>
                <input
                  aria-label="Search"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-mocked-MuiInputBase-input-MuiInput-input"
                  id="mui-2"
                  placeholder="Search"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-mocked-MuiInputAdornment-root"
                >
                  <button
                    aria-label="Clear Search"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-mocked-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      aria-label="clear"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-mocked-MuiSvgIcon-root"
                      data-testid="clear"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root css-15k1nh5"
            />
            <div
              class="MuiBox-root css-p9qzma"
            >
              <div>
                <div
                  style="display: flex;"
                >
                  <span>
                    <div
                      style="display: flex;"
                    >
                      <button
                        aria-label="Refresh Data"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-mocked-MuiButtonBase-root-MuiIconButton-root"
                        data-mui-internal-clone-element="true"
                        tabindex="0"
                        type="button"
                      >
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedInfo MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorInfo MuiButton-root MuiButton-outlined MuiButton-outlinedInfo MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorInfo css-mocked-MuiButtonBase-root-MuiButton-root"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-mocked-MuiButton-startIcon"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                              data-testid="RefreshIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4z"
                              />
                            </svg>
                          </span>
                          Refresh
                        </button>
                        <span
                          class="MuiTouchRipple-root css-mocked-MuiTouchRipple-root"
                        />
                      </button>
                      <button
                        aria-label="Add New RSU"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-mocked-MuiButtonBase-root-MuiIconButton-root"
                        data-mui-internal-clone-element="true"
                        tabindex="0"
                        type="button"
                      >
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-mocked-MuiButtonBase-root-MuiButton-root"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-mocked-MuiButton-startIcon"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                              data-testid="AddCircleOutlineIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8"
                              />
                            </svg>
                          </span>
                          New
                        </button>
                        <span
                          class="MuiTouchRipple-root css-mocked-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-187halg"
          >
            <div>
              <div
                style="overflow-y: auto;"
              >
                <div>
                  <table
                    class="MuiTable-root css-mocked-MuiTable-root"
                    style="table-layout: fixed;"
                  >
                    <thead
                      class="MuiTableHead-root css-mocked-MuiTableHead-root"
                    >
                      <tr
                        class="MuiTableRow-root MuiTableRow-head css-mocked-MuiTableRow-root"
                      >
                        <th
                          class="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingNone MuiTableCell-sizeMedium css-mocked-MuiTableCell-root"
                          scope="col"
                          style="background-color: rgb(255, 255, 255); width: 48px;"
                        >
                          <span
                            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-mocked-MuiButtonBase-root-MuiCheckbox-root"
                          >
                            <input
                              class="PrivateSwitchBase-input css-1m9pwf3"
                              data-indeterminate="false"
                              type="checkbox"
                            />
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                              data-testid="CheckBoxOutlineBlankIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                              />
                            </svg>
                            <span
                              class="MuiTouchRipple-root css-mocked-MuiTouchRipple-root"
                            />
                          </span>
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall css-mocked-MuiTableCell-root"
                          scope="col"
                          style="background-color: rgb(255, 255, 255); box-sizing: border-box;"
                        >
                          <div
                            aria-describedby="rfd-hidden-text-0-hidden-text-0"
                            data-rfd-drag-handle-context-id="0"
                            data-rfd-drag-handle-draggable-id="0"
                            data-rfd-draggable-context-id="0"
                            data-rfd-draggable-id="0"
                            draggable="false"
                            role="button"
                            style="position: relative; min-width: 0; display: flex;"
                            tabindex="0"
                          >
                            <span
                              aria-sort="none"
                              class="MuiButtonBase-root MuiTableSortLabel-root css-mocked-MuiButtonBase-root-MuiTableSortLabel-root"
                              data-testid="mtableheader-sortlabel"
                              role="columnheader"
                              tabindex="0"
                            >
                              Milepost
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-mocked-MuiSvgIcon-root-MuiTableSortLabel-icon"
                                data-testid="arrow_downward"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8z"
                                />
                              </svg>
                            </span>
                            <span
                              data-testid="material-table-column-sort-order-indicator"
                              style="width: 1em;"
                            />
                          </div>
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall css-mocked-MuiTableCell-root"
                          scope="col"
                          style="background-color: rgb(255, 255, 255); box-sizing: border-box;"
                        >
                          <div
                            aria-describedby="rfd-hidden-text-0-hidden-text-0"
                            data-rfd-drag-handle-context-id="0"
                            data-rfd-drag-handle-draggable-id="1"
                            data-rfd-draggable-context-id="0"
                            data-rfd-draggable-id="1"
                            draggable="false"
                            role="button"
                            style="position: relative; min-width: 0; display: flex;"
                            tabindex="0"
                          >
                            <span
                              aria-sort="none"
                              class="MuiButtonBase-root MuiTableSortLabel-root css-mocked-MuiButtonBase-root-MuiTableSortLabel-root"
                              data-testid="mtableheader-sortlabel"
                              role="columnheader"
                              tabindex="0"
                            >
                              IP Address
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-mocked-MuiSvgIcon-root-MuiTableSortLabel-icon"
                                data-testid="arrow_downward"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8z"
                                />
                              </svg>
                            </span>
                            <span
                              data-testid="material-table-column-sort-order-indicator"
                              style="width: 1em;"
                            />
                          </div>
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall css-mocked-MuiTableCell-root"
                          scope="col"
                          style="background-color: rgb(255, 255, 255); box-sizing: border-box;"
                        >
                          <div
                            aria-describedby="rfd-hidden-text-0-hidden-text-0"
                            data-rfd-drag-handle-context-id="0"
                            data-rfd-drag-handle-draggable-id="2"
                            data-rfd-draggable-context-id="0"
                            data-rfd-draggable-id="2"
                            draggable="false"
                            role="button"
                            style="position: relative; min-width: 0; display: flex;"
                            tabindex="0"
                          >
                            <span
                              aria-sort="none"
                              class="MuiButtonBase-root MuiTableSortLabel-root css-mocked-MuiButtonBase-root-MuiTableSortLabel-root"
                              data-testid="mtableheader-sortlabel"
                              role="columnheader"
                              tabindex="0"
                            >
                              Primary Route
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-mocked-MuiSvgIcon-root-MuiTableSortLabel-icon"
                                data-testid="arrow_downward"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8z"
                                />
                              </svg>
                            </span>
                            <span
                              data-testid="material-table-column-sort-order-indicator"
                              style="width: 1em;"
                            />
                          </div>
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall css-mocked-MuiTableCell-root"
                          scope="col"
                          style="background-color: rgb(255, 255, 255); box-sizing: border-box;"
                        >
                          <div
                            aria-describedby="rfd-hidden-text-0-hidden-text-0"
                            data-rfd-drag-handle-context-id="0"
                            data-rfd-drag-handle-draggable-id="3"
                            data-rfd-draggable-context-id="0"
                            data-rfd-draggable-id="3"
                            draggable="false"
                            role="button"
                            style="position: relative; min-width: 0; display: flex;"
                            tabindex="0"
                          >
                            <span
                              aria-sort="none"
                              class="MuiButtonBase-root MuiTableSortLabel-root css-mocked-MuiButtonBase-root-MuiTableSortLabel-root"
                              data-testid="mtableheader-sortlabel"
                              role="columnheader"
                              tabindex="0"
                            >
                              RSU Model
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-mocked-MuiSvgIcon-root-MuiTableSortLabel-icon"
                                data-testid="arrow_downward"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8z"
                                />
                              </svg>
                            </span>
                            <span
                              data-testid="material-table-column-sort-order-indicator"
                              style="width: 1em;"
                            />
                          </div>
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall css-mocked-MuiTableCell-root"
                          scope="col"
                          style="background-color: rgb(255, 255, 255); box-sizing: border-box;"
                        >
                          <div
                            aria-describedby="rfd-hidden-text-0-hidden-text-0"
                            data-rfd-drag-handle-context-id="0"
                            data-rfd-drag-handle-draggable-id="4"
                            data-rfd-draggable-context-id="0"
                            data-rfd-draggable-id="4"
                            draggable="false"
                            role="button"
                            style="position: relative; min-width: 0; display: flex;"
                            tabindex="0"
                          >
                            <span
                              aria-sort="none"
                              class="MuiButtonBase-root MuiTableSortLabel-root css-mocked-MuiButtonBase-root-MuiTableSortLabel-root"
                              data-testid="mtableheader-sortlabel"
                              role="columnheader"
                              tabindex="0"
                            >
                              Serial Number
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-mocked-MuiSvgIcon-root-MuiTableSortLabel-icon"
                                data-testid="arrow_downward"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8z"
                                />
                              </svg>
                            </span>
                            <span
                              data-testid="material-table-column-sort-order-indicator"
                              style="width: 1em;"
                            />
                          </div>
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingCheckbox MuiTableCell-sizeMedium css-mocked-MuiTableCell-root"
                          scope="col"
                          style="text-align: center; background-color: rgb(255, 255, 255); width: 96px; box-sizing: border-box;"
                        >
                          <span
                            aria-disabled="true"
                            class="MuiButtonBase-root Mui-disabled MuiTableSortLabel-root css-mocked-MuiButtonBase-root-MuiTableSortLabel-root"
                            role="button"
                            tabindex="-1"
                          >
                            Actions
                          </span>
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="MuiTableBody-root css-mocked-MuiTableBody-root"
                    >
                      <tr
                        class="MuiTableRow-root css-mocked-MuiTableRow-root"
                        style="height: 245px;"
                      >
                        <td
                          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-mocked-MuiTableCell-root"
                          colspan="7"
                          style="padding-top: 0px; padding-bottom: 0px; text-align: center;"
                        >
                          No records to display
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
          <table
            class="MuiTable-root css-mocked-MuiTable-root"
          >
            <tfoot
              class="MuiTableFooter-root css-mocked-MuiTableFooter-root"
              style="display: grid;"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-footer css-mocked-MuiTableRow-root"
                style="display: grid;"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium MuiTablePagination-root css-mocked-MuiTableCell-root-MuiTablePagination-root"
                  colspan="3"
                >
                  <div
                    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-mocked-MuiToolbar-root-MuiTablePagination-toolbar"
                  >
                    <div
                      class="MuiTablePagination-spacer css-mocked-MuiTablePagination-spacer"
                    />
                    <p
                      class="MuiTablePagination-selectLabel css-mocked-MuiTablePagination-selectLabel"
                      id="mui-11"
                    >
                      Rows per page:
                    </p>
                    <div
                      class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-mocked-MuiInputBase-root-MuiTablePagination-select"
                    >
                      <div
                        aria-controls="mui-5"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-labelledby="mui-11 mui-10"
                        class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-mocked-MuiSelect-select-MuiInputBase-input"
                        id="mui-10"
                        role="combobox"
                        tabindex="0"
                      >
                        <div
                          class="MuiBox-root css-c8mw91"
                        >
                          5 rows 
                        </div>
                      </div>
                      <input
                        aria-hidden="true"
                        class="MuiSelect-nativeInput css-mocked-MuiSelect-nativeInput"
                        tabindex="-1"
                        value="5"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-mocked-MuiSvgIcon-root-MuiSelect-icon"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                    </div>
                    <p
                      class="MuiTablePagination-displayedRows css-mocked-MuiTablePagination-displayedRows"
                    >
                      0â€“0 of 0
                    </p>
                    <div
                      class="MuiBox-root css-lm78jv"
                    >
                      <span
                        aria-label="First Page"
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <button
                          aria-label="First Page"
                          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-mocked-MuiButtonBase-root-MuiIconButton-root"
                          disabled=""
                          tabindex="-1"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                            data-testid="first_page"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M18.41 16.59 13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
                            />
                          </svg>
                        </button>
                      </span>
                      <span
                        aria-label="Previous Page"
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <button
                          aria-label="Previous Page"
                          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-mocked-MuiButtonBase-root-MuiIconButton-root"
                          disabled=""
                          tabindex="-1"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                            data-testid="chevron_left"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                            />
                          </svg>
                        </button>
                      </span>
                      <span
                        class="MuiTypography-root MuiTypography-caption css-mocked-MuiTypography-root"
                        style="flex: 1; text-align: center; align-self: center; flex-basis: inherit;"
                      >
                        0-0 of 0
                      </span>
                      <span
                        aria-label="Next Page"
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <button
                          aria-label="Next Page"
                          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-mocked-MuiButtonBase-root-MuiIconButton-root"
                          disabled=""
                          tabindex="-1"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                            data-testid="chevron_right"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                            />
                          </svg>
                        </button>
                      </span>
                      <span
                        aria-label="Last Page"
                        class=""
                        data-mui-internal-clone-element="true"
                      >
                        <button
                          aria-label="Last Page"
                          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-mocked-MuiButtonBase-root-MuiIconButton-root"
                          disabled=""
                          tabindex="-1"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-mocked-MuiSvgIcon-root"
                            data-testid="last_page"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M5.59 7.41 10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
                            />
                          </svg>
                        </button>
                      </span>
                    </div>
                  </div>
                </td>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
